### PR TITLE
Refactored userRegister: controller, service and tests

### PR DESCRIPTION
**Pull request by diegdar**

### CODE REVIEW

File: app/Service/User/UserRegisterService.php [hunk 1]
```diff
@@ -7,43 +7,71 @@
...
 use App\Models\User;
-use Illuminate\Support\Facades\DB;
 use Exception;
 use PDOException;
+use Error;
 
 class UserRegisterService
 {
-    public function createUser(array $input): array | bool
+    public function createUser(array $input): array
```
Comment (jordimorillo) [inline app/Service/User/UserRegisterService.php:16]: use "execute(array $input)" instead

File: app/Service/User/UserRegisterService.php [hunk 2]
```diff
@@ -7,43 +7,71 @@
...
+        return $success;
+    }
+
+    private function createUserRecord(array $input): User
+    {
         try {
-            $user = User::create($input);
+            return User::create($input);
+        } catch (PDOException $e) {
+            throw new Exception('Error al crear el usuario en la base de datos: ' . $e->getMessage());
```
Comment (jordimorillo) [inline app/Service/User/UserRegisterService.php:35]: Make a UserRegisterException class to handle this error (extending from DomainException)

File: app/Service/User/UserRegisterService.php [hunk 3]
```diff
@@ -7,43 +7,71 @@
...
+        } catch (Error $e) {
+            throw new Exception('Error inesperado al crear el usuario: ' . $e->getMessage());
```
Comment (jordimorillo) [inline app/Service/User/UserRegisterService.php:37]: You don't need two catches for this, just take the error and return UserRegisterException for both of them.

File: app/Http/Controllers/api/RegisterController.php [hunk 4]
```diff
@@ -7,57 +8,34 @@
...
-                'exception' => $e->getMessage(),
-                'trace' => $e->getTraceAsString(),
-            ]);
-            return response()->json('Database error occurred during registration.', 500);
+            $result = $this->userRegisterService->createUser($input);
+
+            DB::commit();
+
+            return response()->json($result, 200);
         } catch (Exception $e) {
```
Comment (jordimorillo) [inline app/Http/Controllers/api/RegisterController.php:36]: Specify the catch for \DomainException cases

Comment (jordimorillo) [reply app/Http/Controllers/api/RegisterController.php:36]: Cuando es \DomainException, se supone que estamos gestionando el error nosotros. Por tanto, viene con un "$e->getCode()" que hemos preparado para que coincida con un error de tipo HTTP. Por tanto, se puede hacer esto:

```php
catch (\DomainException $e) {
            DB::rollBack();
            Log::error('Domain exception:', [
                'exception' => $e->getMessage(),
                'trace' => $e->getTraceAsString(),
            ]);
            return response()->json(null, $e->getCode()); // We want the error is only shown in log report
        }
```

File: app/Service/User/UserRegisterService.php [hunk 5]
```diff
@@ -7,43 +7,71 @@
...
 
+    private function createStudentRecord(string $userId): Student
+    {
+        try {
             $student = new Student;
-            $student->user_id = $user->id;
+            $student->user_id = $userId;
             $student->save();
+            return $student;
+        } catch (PDOException $e) {
```
Comment (jordimorillo) [inline app/Service/User/UserRegisterService.php:48]: Specify this exception

File: app/Service/User/UserRegisterService.php [hunk 6]
```diff
@@ -7,43 +7,71 @@
...
+            throw new Exception('Error al crear el estudiante en la base de datos: ' . $e->getMessage());
+        } catch (Error $e) {
```
Comment (jordimorillo) [inline app/Service/User/UserRegisterService.php:50]: The same

Comment (jordimorillo) [reply app/Service/User/UserRegisterService.php:50]: Review next Exception and specify them

File: app/Service/User/UserRegisterService.php [hunk 7]
```diff
@@ -7,43 +7,71 @@
...
-            return $success;
-        } catch (PDOException | Exception $e) {
-            DB::rollBack();
-            return false;
+    private function generateAccessToken(User $user): string
+    {
+        try {
+            return $user->createToken('ITAcademy')->accessToken;
+        } catch (Exception $e) {
+            throw new Exception('Error al generar el token de acceso: ' . $e->getMessage());
```
Comment (jordimorillo) [inline app/Service/User/UserRegisterService.php:74]: TokenGenerateException

File: tests/Feature/Service/User/RegisterUserServiceTest.php [hunk 8]
```diff
@@ -36,65 +36,25 @@ private function createUserData(): array
...
-    {
-        $userData = $this->createUserData();
-        $user = User::create($userData);
-        $student = Student::create(['user_id' => $user->id]);
-
-        $this->assertNotEmpty($student->id);
-    }
-
-    public function test_resume_uuid_is_generated_when_user_is_created(): void
+    public function test_UserRegisterService_can_be_instantiated(): void
```
Comment (jordimorillo) [inline tests/Feature/Service/User/RegisterUserServiceTest.php:39]: never use snake case in tests. we use camel case

File: tests/Feature/Service/User/RegisterUserServiceTest.php [hunk 9]
```diff
@@ -124,10 +83,8 @@ public function test_user_creation_with_empty_data(): void
             'password' => '',
         ];
 
-        $success = $this->userService->createUser($registerData);
-
-        $this->assertFalse($success);//Assert that the user creation was unsuccessful
-
+        $this->expectException(\Exception::class);
```
Comment (jordimorillo) [inline tests/Feature/Service/User/RegisterUserServiceTest.php:86]: Expect a UserCreateException here

File: tests/Feature/Service/User/RegisterUserServiceTest.php [hunk 10]
```diff
@@ -138,15 +95,13 @@ public function test_user_creation_with_empty_data(): void
...
         if ($resultCorrect) {
-            $this->assertNotEmpty($success['email']);
-            $this->assertNotEmpty($success['token']);
+            $response = $this->userService->createUser($array);
+            $this->assertNotEmpty($response['email']);
+            $this->assertNotEmpty($response['token']);
         } else {
-            // If the result is expected to be incorrect, assert that the creation was unsuccessful
-            $this->assertFalse($success);
+            $this->expectException(\Exception::class);
```
Comment (jordimorillo) [inline tests/Feature/Service/User/RegisterUserServiceTest.php:103]: Same

**Review by jordimorillo — COMMENTED** [review]
Your code is very good but take a look at this.

File: app/Service/User/UserRegisterService.php (hunk 2 above)
Comment (diegdar) [reply app/Service/User/UserRegisterService.php:35]: done

File: app/Service/User/UserRegisterService.php (hunk 3 above)
Comment (diegdar) [reply app/Service/User/UserRegisterService.php:37]: done

File: app/Service/User/UserRegisterService.php (hunk 5 above)
Comment (diegdar) [reply app/Service/User/UserRegisterService.php:48]: done

File: app/Service/User/UserRegisterService.php [hunk 11]
```diff
@@ -7,43 +7,71 @@
...
+    {
+        try {
             $resume = new Resume;
-            $resume->student_id = $student->id;
-            $resume->specialization = $input['specialization'] ?? null;
+            $resume->student_id = $studentId;
+            $resume->specialization = $specialization;
             $resume->save();
+        } catch (PDOException $e) {
+            throw new Exception('Error al crear el resume en la base de datos: ' . $e->getMessage());
```
[the remark this answers is not in the document]
Comment (diegdar) [reply app/Service/User/UserRegisterService.php:63]: done

File: app/Service/User/UserRegisterService.php (hunk 7 above)
Comment (diegdar) [reply app/Service/User/UserRegisterService.php:74]: done

File: app/Http/Controllers/api/RegisterController.php (hunk 4 above)
Comment (diegdar) [reply app/Http/Controllers/api/RegisterController.php:36]: done

File: tests/Feature/Service/User/RegisterUserServiceTest.php (hunk 9 above)
Comment (diegdar) [reply tests/Feature/Service/User/RegisterUserServiceTest.php:86]: done

**Review by JustTheDude001 — APPROVED** [review]
# Pull Request feature/signup-endpoint#136

# Checklist:
- [x] Git Diff For Change Checking
- [x] Tests - php artisan test
- [x] Check Code Files
	- [x]  app/Exceptions/ResumeCreateException.php           |   14 +
 	- [x] app/Exceptions/TokenGenerateException.php          |   14 +
 	- [x] app/Exceptions/UserRegisterException.php           |   22 +
 	- [x] app/Http/Controllers/api/AuthController.php        |   92 +-
 	- [x] app/Http/Controllers/api/RegisterController.php    |   50 +-
 	- [x] app/Http/Requests/RegisterRequest.php              |    1 +
 	- [ ] app/Service/User/UserRegisterService.php           |   88 +-
 	- [ ] composer.lock                                      |   63 +-
 	- [ ] docker-compose.yml                                 |   15 +-
 	- [x] routes/api/v1.php                                  |    4 +-
 	- [x] .../Controller/User/RegisterUserControllerTest.php |   47 +-
 	- [x] .../Service/User/RegisterUserServiceTest.php  



- git diff --stat main | grep +
```
git diff --stat main | grep +
 app/Exceptions/ResumeCreateException.php           |   14 +
 app/Exceptions/TokenGenerateException.php          |   14 +
 app/Exceptions/UserRegisterException.php           |   22 +
 app/Http/Controllers/api/AuthController.php        |   92 +-
 app/Http/Controllers/api/RegisterController.php    |   50 +-
 app/Http/Requests/RegisterRequest.php              |    1 +
 app/Service/User/UserRegisterService.php           |   88 +-
 composer.lock                                      |   63 +-
 docker-compose.yml                                 |   15 +-
 routes/api/v1.php                                  |    4 +-
 .../Controller/User/RegisterUserControllerTest.php |   47 +-
 .../Service/User/RegisterUserServiceTest.php       |   95 +-
 406 files changed, 234 insertions(+), 2511 deletions(-)

```

## Tests:
image


## Corrections:
- Missing Strict Types in ALL new Exceptions:
	+ ResumeCreateException.php
	+ TokenGenerateException.php
	+ UserRegisterException.php
	
	*Check File LanguageNotFoundException.php*
	
- Missing Strict Types in AuthController.php AND therefore "typing" the functions login, verifyUser and logout (the output)

- 



## Notes
- Maybe there are troubles while merging the files:  
	+ composer.lock
	+ docker-compose.yml

- Maybe there are troubles in the future with the token generation to login as user. As there are different methods for register and signin. Nothing to do now, just a warning.

### Congrats!!! Great job!!! I accept the PR, do not forget to add types stricts to the missing files if needed... I do not really know if it is persuatory, but it helped me in the past to fix some code ^^

**Review by begoIT24 — APPROVED** [review]
@JustTheDude001 has done a great job reviewing this. It seems right for me too.
I have tried the testing and it passed Ok.
image
Well done @diegdar

**Pull request merged**